[ENH] benchmarks: add tqdm progress bar

Spinachboul opened the pull request:
#### Reference Issues/PRs
Fixes Issue https://github.com/sktime/sktime/issues/6560


#### What does this implement/fix? Explain your changes.
Adds progress bar using tqdm library 

#### What should a reviewer concentrate their feedback on?
Focus on testing the progress while calling the API

#### Did you add any tests for the change?

Yes, tested the progress bar functionality using sample model registry and model validation

![Capture](https://github.com/sktime/sktime/assets/105979087/42a09c29-6af2-4518-81b4-b6c05c2631d7)

yarnabrina commented on sktime/benchmarking/_lib_mini_kotsu/run.py:
```diff
@@ -6,6 +6,7 @@
 from typing import List, Optional, Union
 
 import pandas as pd
+from tqdm import tqdm
```
Is `tqdm` part of our core dependencies, or implied by those? If not, this is going to fail.

Spinachboul replied:
I'm afraid tqdm is an independent package

yarnabrina replied:
I'm not soubting that, but it could be implied. For example, `statsforecast` (part of `forecasting` subset) will imply `tqdm` and so on.

As far as I know none of the core dependencies, e.g. `pandas`, `numpy`, `scikit-learn` etc. will imply `tqdm`, so this line will fail CI. What you should do is move this import to inside your functions/methods and use `tqdm` only if it present. There is no need to make progress bars mandatory, especially if someone uses it in non-interactive setup.

Spinachboul replied:
@yarnabrina
What dependency do u suggest using for adding the progress bar

fkiraly reviewed (changes requested):
I think this PR makes more substantial changes than just adding a progress bar? For instance, it changes the logger files or the structure of the code.

Change requests:

* leave the logic unchanged in this PR. If you want to change or improve, you are much welcome, but it should be a separate PR.
* I would recommend a refactor: factor out the loop into a single private function, that is, a single execution of `validation_spec` and `model_spec`. This will play well with parallelization efforts later.
* we need to make sure that use of `tqdm` is optional, and the code still runs if `tqdm` is not installed.

Spinachboul commented:
@fkiraly 
Sure then! I'll keep the base logic of the code unchanged , but just add the progress bars in the console 
Could I import the dependencies inside the functions to avoid CI errors during checks!

fkiraly commented:
> Could I import the dependencies inside the functions to avoid CI errors during checks!

Yes, I would also make sure that everything runs without the `tqdm` dependency as well.

Spinachboul commented:
@fkiraly & @benHeid 
I hope this should be the final commit before merging!
Please tell if any changes need to be made to the code.

fkiraly reviewed (changes requested):
Thanks!

I think there should be verbose mode and non-verbose mode, I would switch that with a `verbose` arg that can be True or False (default). If False, `tqdm` is not imported, and no prints are done. 

I think this is quite easy to achieve, since you can do an easy switch on the generator used in the two `for` loops, e.g., `if verbose: model_registry_generator = tqdm(...)` etc

There should also be some consistency with the `logger` use.

Spinachboul commented:
@fkiraly & @benHeid 
Would this be to your liking
I have included a verbose as one of the params which will control the verbosity of the progress bars
If `verbose = False`, ,then the funtion `tqdm` defined would let the program remain functional else would invoke progress bars while executing the **for loops**
Also I have made the program consistent by using `logger.info()` instead of `print` statements

fkiraly commented:
> Would this be to your liking

This works but imo it is unnecessarily obscure. How about

```python
if verbose and tqdm_available:
    validation_gen = tqdm(validation_registry.all())
else:
    validation_gen = validation_registry.all()
```

yarnabrina commented:
I'm unable to see any conditions in the changes tab on this Github PR currently regarding verbose mode. I'm also unable to see any check for presence of tqdm soft dependency being installed or not.

Is it just me? @fkiraly are you able to see?

benHeid commented on sktime/benchmarking/_lib_mini_kotsu/run.py:
```diff
@@ -58,16 +59,22 @@ def run(
-    for validation_spec in validation_registry.all():
+    print("Printing Validations...")
+    print("\n")
+    for validation_spec in tqdm(validation_registry.all()):
```
I would propose to avoid the print statements, you can add a description of the loop into the tqdm call.

benHeid commented on sktime/benchmarking/_lib_mini_kotsu/run.py:
```diff
@@ -58,16 +59,22 @@ def run(
-        for model_spec in model_registry.all():
+        print("Running models...")
+        print("\n")
+        for model_spec in tqdm(model_registry.all()):
```
I would propose to set `leave=False` to the second loop. This would remove the inner progress bar after the loop is finished. This would avoid a wall of loops and the user has not to scroll up to see the outer loop.

benHeid reviewed (changes requested):
Thank you for the contribution. I have two requests that would improve the user experience.

fkiraly commented:
> I'm unable to see any conditions in the changes tab on this Github PR currently regarding verbose mode. I'm also unable to see any check for presence of tqdm soft dependency being installed or not.

I think the changes are only in the screenshot, which does not correspond ot the state of the fork. @Spinachboul, can you please make sure you update the PR to the most recent state?

Spinachboul commented:
@fkiraly | @benHeid  | @yarnabrina | @achieveordie 
I have updated the PR in the following ways

- Added the verbosity to the progress bars as suggested by @fkiraly 
- Removed all the `print` statements and addional `logger info()`, while adding desciptions to the progress bars as suggested by @benHeid 
- Also removed the code obscurity by importing `tqdm` additionally as suggested by @fkiraly 
- Even updated the params in the `main` function!

fkiraly commented on sktime/benchmarking/_lib_mini_kotsu/run.py:
```diff
@@ -1,4 +1,3 @@
-"""Interface for running a registry of models on a registry of validations."""
```
why did you remove this line?

fkiraly commented on sktime/benchmarking/_lib_mini_kotsu/run.py:
```diff
@@ -58,16 +61,42 @@ def run(
 
+    tqdm_available = False
+    if verbose:
+        try:
```
try/except is not good style, if you know the condition you want to check.
In that case, simply check the condition, you can do this with `_check_soft_dependencies` as I have recommended, kindly search the code base.

fkiraly commented on sktime/benchmarking/_lib_mini_kotsu/run.py:
```diff
@@ -79,15 +108,11 @@ def run(
                 logger.info(
                     f"Skipping validation - model: "
-                    f"{validation_spec.id} - {model_spec.id}"
-                    ", as found prior result in results."
```
why are you removing the comma?

fkiraly commented on sktime/benchmarking/_lib_mini_kotsu/run.py:
```diff
@@ -79,15 +108,11 @@ def run(
                 )
                 continue
 
-            logger.info(
```
why are you removing this?

fkiraly reviewed (changes requested):
Great! Left some comments above.

Spinachboul commented:
Sure!
I'll fix them soon
Thanks for the review

Spinachboul commented:
@fkiraly 
Made the requested changes!
Please review

fkiraly commented on sktime/benchmarking/_lib_mini_kotsu/run.py:
```diff
@@ -58,16 +65,37 @@ def run(
         results_df = pd.DataFrame(columns=["validation_id", "model_id", "runtime_secs"])
         results_df["runtime_secs"] = results_df["runtime_secs"].astype(int)
 
+    tqdm_available, _ = _check_soft_dependencies("tqdm", severity="warning")
```
I think the warning should be raised only if `verbose`, with a clear error message. Otherwise, I would just leave it silent.

fkiraly replied:
If you want to warn, you should use `warn` from `sktime.utils.warnings` which receives a reference from `self`, so users can turn it off via `set_config`.

Spinachboul replied:
@fkiraly 
```{python}
     try:
        results_df = pd.read_csv(results_path)
    except FileNotFoundError:
        results_df = pd.DataFrame(columns=["validation_id", "model_id", "runtime_secs"])
        results_df["runtime_secs"] = results_df["runtime_secs"].astype(int)

    tqdm_available, _ = _check_soft_dependencies("tqdm", severity="none")

    if verbose and not tqdm_available:
        warn("tqdm is not installed. Continuing without progress bars.", logger)

    if tqdm_available:
        from tqdm import tqdm
```
Something like this ⬆ ?

**This has not been committed!**

fkiraly replied:
yes, though use `from sktime.utils.warnings import warn`, there you can pass the `self` reference.

fkiraly commented on sktime/forecasting/hf_transformers_forecaster.py:
```diff
@@ -57,32 +59,6 @@ class HFTransformersForecaster(BaseForecaster):
         Whether the predictions should be deterministic or not.
     callbacks : list, default=[]
         List of callbacks to use during training. See `transformers.Trainer`
-
```
why delete the examples?

fkiraly reviewed (changes requested):
You are making changes in files unrelated to the benchmark.

I would guess you rebased or otherwise corrupted your fork, kindly ensure that no changes are made to files unrelated to your intended change, e.g., `hf_transformers_forecaster`

Spinachboul commented:
@fkiraly 
Why is the check **Install and test / code-quality (pull_request)** failing for this PR??

benHeid commented:
The imports aren't sorted correctly. Thus, isort is failing. Please execute isort locally and push again

fkiraly commented:
Here's the guide for how to ensure formatted code: https://www.sktime.net/en/latest/developer_guide/coding_standards.html

Spinachboul commented:
@fkiraly 
The commits have been messed for this PR
Can I close this one and open another one with the changes to the code file??
This way it would be easier to fix things up

fkiraly commented:
I think it's easy to fix - simply copy the `hf_transformers_forecaster` from main and commit it.

Your commits will be squashed anyway into a single one, so the changes to that file will not show.

Spinachboul commented:
I'll be closing this PR since I would be opening another one which solves a new issue
Ofc I'll be including this commit also in the new PR
Hope that is fine by you!
@fkiraly & @benHeid
